src/consensus/block_producer.rs

```diff
@@ -240,96 +279,30 @@ where
                             debug!("parent is ready, continuing with same parent");
                         }
                         // ParentReady was seen, start the DELTA_BLOCK timer
-                        // account for the time it took to finish producing the slice
+                        // account for the time it took to finish producing current slice
                         debug!("starting blocktime timer");
-                        let elapsed = Instant::now() - start;
-                        let duration = self.delta_block.saturating_sub(elapsed);
-                        (payload, duration)
-                  }
+                        let time_left = self.delta_block.saturating_sub(start.elapsed());
+                        (payload, time_left)
+                    }
                 }
             };
 
             let is_last = slice_index.is_max() || new_duration_left.is_zero();
-            if is_last && !parent_ready_receiver.is_terminated() {
-                let (new_slot, new_hash) = (&mut parent_ready_receiver).await.unwrap();
-                if new_hash != parent_hash {
-                    assert_ne!(new_slot, parent_slot);
-                    debug!(
-                        "changed parent from {} in slot {} to {} in slot {}",
-                        &hex::encode(parent_hash)[..8],
-                        parent_slot,
-                        &hex::encode(new_hash)[..8],
-                        new_slot
-                    );
-                    payload.parent = Some((new_slot, new_hash));
-                } else {
-                    debug!("parent is ready, continuing with same parent");
-                }
-            }
-            let header = SliceHeader {
-                slot,
-                slice_index,
-                is_last,
-            };
-
-            match self.shred_and_disseminate(header, payload).await? {
-                Some(block_hash) => return Ok((slot, block_hash)),
-                None => {
-                    assert!(!new_duration_left.is_zero());
-                    duration_left = new_duration_left;
-                }
+            if let Some(ref parent_rcv) = parent_ready_receiver
+                && is_last
+            {
+                assert!(parent_rcv.is_terminated());
```

Comment:
I am not sure if I follow the logic that this condition should always hold.  slice_index.is_max() can be true without us seeing the parent ready event, no?